File: src/main/java/com/aoranzhang/ezrentback/data/graphql/dataTypes/GraphqlDateType.java
package com.aoranzhang.ezrentback.data.graphql.dataTypes;

import graphql.language.StringValue;
import graphql.schema.Coercing;
import graphql.schema.GraphQLScalarType;
import org.springframework.stereotype.Component;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

@Component
public class GraphqlDateType extends GraphQLScalarType {
    public GraphqlDateType() {

        super("Date", "Date", new Coercing<Date, String>() {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd kk:mm");

            @Override
            public String serialize(Object input) {
                if(input instanceof Date) {
                    return sdf.format(input);
                }else {
                    return null;
                }
            }

            @Override
            public Date parseValue(Object input) {
                if(input instanceof String) {
                    try {
                        return sdf.parse((String) input);
                    }catch (ParseException e) {
                        return null;
                    }
                }else {
                    return null;
                }
            }

            @Override
            public Date parseLiteral (Object input) {
                if (input instanceof StringValue) {
                    try {
                        return sdf.parse(((StringValue) input).getValue());
                    } catch (ParseException e) {
                        return null;
                    }
                }
                return null;
            }
        });
    }
}

File: src/main/kotlin/com/aoranzhang/ezrentback/data/graphql/dataTypes/GraphqlDateType.kt
package com.aoranzhang.ezrentback.data.graphql.dataTypes

import graphql.language.StringValue
import graphql.schema.Coercing
import graphql.schema.GraphQLScalarType
import org.springframework.stereotype.Component

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date

@Component
class GraphqlDateType : GraphQLScalarType("Date", "Date", object : Coercing<Date, String> {
    var sdf = SimpleDateFormat("yyyy-MM-dd kk:mm")

    override fun serialize(input: Any): String? {
        return if (input is Date) {
            sdf.format(input)
        } else {
            null
        }
    }

    override fun parseValue(input: Any): Date? {
        return if (input is String) {
            try {
                sdf.parse(input)
            } catch (e: ParseException) {
                null
            }
        } else {
            null
        }
    }

    override fun parseLiteral(input: Any): Date? {
        if (input is StringValue) {
            try {
                return sdf.parse(input.getValue())
            } catch (e: ParseException) {
                return null
            }

        }
        return null
    }
})
